### PR TITLE
add diox

### PR DESCRIPTION
**Pull request by bstolinski**
## :scroll: Description

Currently [dio](https://github.com/flutterchina/dio) library is very poorly maintained. An alternative could be [diox](https://github.com/cfug/diox/issues/29) which is a fork of dio. They have currently fixed many [problems](https://github.com/orgs/cfug/projects/1) that have been unresolved for months on the original repository. 

## :green_heart: How did you test it?
Unit tests

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [ ] No new PII added or SDK only sends newly added PII if `sendDefaultPii` is enabled
- [x] I updated the docs if needed
- [x] All tests passing
- [x] No breaking changes

### CODE REVIEW

**Comment by ueman** [comment]
Hello, I'm one of the devs behind diox and also happen to have written the original Sentry dio integration. Thanks for doing this :D

I believe you can try to remove the ignores for the `strict_raw_type` lint. Some should be fixed, but not all. So that should be at least double-checked.

The dio event processor was reworked quite a bit on the v7 branch. So depending on the timeline for v7 (@marandaneto knows more) you might want to take the v7 code of the dio integration and then also target the v7 branch.

CI is also completely missing.

File: diox/lib/sentry_dio.dart [hunk 1]
```diff
@@ -0,0 +1,7 @@
+library sentry_diox;
```
Comment (ueman) [inline diox/lib/sentry_dio.dart]: This file should be named `sentry_diox.dart`

File: diox/lib/src/breadcrumb_client_adapter.dart [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+import 'package:diox/io.dart';
+import 'package:sentry/sentry.dart';
+
+/// A [Diox](https://pub.dev/packages/diox)-package compatible HTTP client adapter
+/// which records requests as breadcrumbs.
+///
+/// Remarks:
+/// If this client is used as a wrapper, a call to close also closes the
+/// given client.
+class BreadcrumbClientAdapter extends IOHttpClientAdapter {
```
Comment (ueman) [inline diox/lib/src/breadcrumb_client_adapter.dart]:
```suggestion
class BreadcrumbClientAdapter implements HttpClientAdapter {
```

File: diox/lib/src/breadcrumb_client_adapter.dart [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  // ignore: public_member_api_docs
```
Comment (ueman) [inline diox/lib/src/breadcrumb_client_adapter.dart:13]: This should be documented

File: diox/lib/src/sentry_diox_client_adapter.dart [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+/// It starts and finishes a Span if there's a transaction bound to the Scope
+/// through the [TracingClientAdapter] client, it's on by default.
+///
+/// Remarks: If this client is used as a wrapper, a call to close also closes
+/// the given client.
+///
+/// Remarks:
+/// HTTP traffic can contain PII (personal identifiable information).
+/// Read more on data scrubbing [here](https://docs.sentry.io/product/data-management-settings/advanced-datascrubbing/).
+class SentryDioxClientAdapter extends IOHttpClientAdapter {
```
Comment (ueman) [inline diox/lib/src/sentry_diox_client_adapter.dart]: Same thing here as for the other adapter

File: diox/lib/src/tracing_client_adapter.dart [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+import 'dart:typed_data';
+
+import 'package:diox/diox.dart';
+import 'package:diox/io.dart';
+import 'package:sentry/sentry.dart';
+
+/// A [Diox](https://pub.dev/packages/diox)-package compatible HTTP client adapter
+/// which adds support to Sentry Performance feature.
+/// https://develop.sentry.dev/sdk/performance
+class TracingClientAdapter extends IOHttpClientAdapter {
```
Comment (ueman) [inline diox/lib/src/tracing_client_adapter.dart]: Same thing here as for the other adapter

File: diox/pubspec.yaml [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+version: 6.19.0
+homepage: https://docs.sentry.io/platforms/dart/
+repository: https://github.com/getsentry/sentry-dart
+issue_tracker: https://github.com/getsentry/sentry-dart/issues
+
+environment:
+  sdk: '>=2.12.0 <3.0.0'
+
+dependencies:
+  diox: ^5.0.0-dev.3
```
Comment (ueman) [inline diox/pubspec.yaml]: If possible, please use 
```suggestion
  diox: ^5.0.0-0
```
That should make it compatible with all v5 version and stable v5 versions of diox.

File: diox/pubspec.yaml [hunk 7]
```diff
@@ -0,0 +1,20 @@
+name: sentry_diox
+description: An integration which adds support for performance tracing for the Diox package.
+version: 6.19.0
+homepage: https://docs.sentry.io/platforms/dart/
+repository: https://github.com/getsentry/sentry-dart
+issue_tracker: https://github.com/getsentry/sentry-dart/issues
+
+environment:
+  sdk: '>=2.12.0 <3.0.0'
```
Comment (ueman) [inline diox/pubspec.yaml]: diox requires Dart >=2.13.0 so there's no point in using Dart 2.12 as min here. Raising the min Dart version allows us to use newer features which we probably want to use if we need them.

File: diox/README.md [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+- Follow the installing instructions on [pub.dev](https://pub.dev/packages/sentry/install).
+
+- Initialize the Sentry SDK using the DSN issued by Sentry.io.
+
+- Call `dio.addSentry()`. This *must* be the last initialization step of the Dio setup, otherwise your configuration of Dio might overwrite the Sentry configuration.
+
+```dart
+import 'package:sentry/sentry.dart';
+import 'package:sentry_dio/sentry_dio.dart';
```
Comment (ueman) [inline diox/README.md]: It should import diox

**Comment by bstolinski** [comment]
Hi @ueman thank you for your efforts!
I have made some of the changes you suggested. I have added you as a collaborator so feel free to make changes as well.

**Comment by marandaneto** [comment]
Hey @bstolinski thank you for the contribution.
I will review this next week.

**Comment by marandaneto** [comment]
So I've checked with my team and we've decided to not support `diox` for now.
Adding new integrations adds the overhead of maintenance on the team, which most of the time is fine when the integration is long-term support and adoption is growing, it's not the case right now since it's still a preview.

Once it has a few stable releases and we can see an adoption trend we can re-evaluate, I'd leave it open, we'll likely not run into conflicts in the future, it's a new package anyway.

To avoid this, please raise a new issue or discussion asking if this is fine and what we think about new integrations or PRs that are a little bit bigger, so we double-check before putting the work in.

Thank you for your contribution.

**Comment by bstolinski** [comment]
Hi @marandaneto, thanks for your reply.
I fully understand your decision, I'll just add that I'll probably keep this fork as up to date as possible because I need it for an application I'm working on.

File: diox/lib/src/diox_event_processor.dart [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+      return exceptions;
+    }
+    var dioSentryException = dioSentryExceptions.first;
+
+    final exceptionIndex = exceptions.indexOf(dioSentryException);
+    exceptions.remove(dioSentryException);
+
+    // Remove error and stacktrace, so that the DioError value doesn't
+    // include the chained exception.
+    final effectiveDioError = DioError(
```
Comment (kuhnroyal) [inline diox/lib/src/diox_event_processor.dart:108]: This is different from the handling in sentry_dio. This `effectiveDioError` needs to be set into `event.throwable` to replace the original error.

**Comment by AlexV525** [comment]
`dio` is back online again.

**Comment by bstolinski** [comment]
> `dio` is back online again.

I am closing this PR, please use https://github.com/getsentry/sentry-dart/tree/main/dio

**Comment by AlexV525** [comment]
There are still a few breaking changes that need to migrate, I would suggest updating the previous integration according to this PR.

**Comment by bstolinski** [comment]
I will try to do this when I have free time.

**Pull request closed without merging**